plugins/jobs/src/js/components/form/helpers/MetronomeJobValidators.ts

```diff
@@ -1,64 +1,108 @@
+  validate(formData: JobOutput): FormError[] {
+    const { run } = formData.job;
+    const parameters = (run.docker && run.docker.parameters) || [];
+
+    // prettier-ignore
+    return pipe(
+      // IS BOOLEAN
+
+      isBoolean(_ => "job.run.docker.forcePullImage", [run.docker && run.docker.forcePullImage]),
+      isBoolean(_ => "job.run.docker.privileged", [run.docker && run.docker.privileged]),
+      isBoolean(_ => "job.run.ucr.privileged", [run.ucr && run.ucr.privileged]),
+      isBoolean(_ => "job.run.ucr.image.forcePull", [run.ucr && run.ucr.image&& run.ucr.image.forcePull]),
+
+      // IS NUMBER
+
+      isNumber(_ => "job.run.cpus", [run.cpus]),
+      isNumber(_ => "job.run.disk", [run.disk]),
+      isNumber(_ => "job.run.gpus", [run.gpus]),
+      isNumber(_ => "job.run.maxLaunchDelay", [run.maxLaunchDelay]),
+      isNumber(_ => "job.run.mem", [run.mem]),
+      isNumber(_ => "job.run.restart.activeDeadlineSeconds", [run.restart && run.restart.activeDeadlineSeconds]),
+      isNumber(_ => "job.run.taskKillGracePeriodSeconds", [run.taskKillGracePeriodSeconds]),
+
+      // IS OBJECT
+
+      isObject(_ => "job.labels", [formData.job.labels]),
+      isObject(_ => "job.run.docker", [run.docker]),
+      isObject(_ => "job.run.ucr", [run.ucr]),
+      isObject(_ => "job.run.ucr.image", [run.ucr && run.ucr.image]),
+
+      // IS PRESENT
+
+      isPresent(_ => "job.id", [formData.job.id]),
+      isPresent(_ => "job.run.cpus", [run.cpus]),
+      isPresent(_ => "job.run.disk", [run.disk]),
+      isPresent(_ => "job.run.mem", [run.mem]),
+      isPresent(i => `job.labels.${i}.key`, Object.keys(formData.job.labels || [])),
+      isPresent(i => `job.run.artifacts.${i}.uri`, (run.artifacts || []).map(_ => _.uri)),
+      isPresent(i => `job.run.docker.parameters.${i}.key`, parameters.map(_ => _.key)),
+      isPresent(i => `job.run.docker.parameters.${i}.value`, parameters.map(_ => _.value)),
+
+      // IS STRING
+
+      isString(_ => "job.id", [formData.job.id]),
+      isString(_ => "job.run.cmd", [run.cmd]),
+      isString(_ => "job.run.docker.image", [run.docker && run.docker.image]),
+      isString(_ => "job.run.restart.policy", [run.restart && run.restart.policy]),
+      isString(_ => "job.run.ucr.image.id", [run.ucr && run.ucr.image && run.ucr.image.id]),
+      isString(_ => "job.run.user", [run.user]),
+      isString(i => `job.labels.${i}.key`, Object.keys(formData.job.labels || [])),
+      isString(i => `job.labels.${i}.value`, Object.values(formData.job.labels || [])),
+      isString(i => `job.run.args.${i}`, run.args || []),
+      isString(i => `job.run.artifacts.${i}.uri`, (run.artifacts || []).map(_ => _.uri)),
+      isString(i => `job.run.docker.parameters.${i}.key`, parameters.map(_ => _.key)),
+      isString(i => `job.run.docker.parameters.${i}.value`, parameters.map(_ => _.value))
+
+      // pipe only infers 10 steps, so we need a cast here
```

Comment:
Can we use more then one pipe?

Reply:
We could, but not casting the type does not give us too much information, so let's forget about it 🤷‍♂️